src/parser/src/UnaryExprNode.cpp

```diff
@@ -1,97 +1,72 @@
-    }
+    auto operandMatNode = std::dynamic_pointer_cast<AST::ExprNode>(this->operand);
+    llvm::Value* matAlloc;
+    if (auto* operandType = std::get_if<Typing::MatrixType>(&*operandMatNode->type)) {
+        matAlloc = Utils::createMatrix(context, *operandType);
+
+        // We generate the operations sequentially
+        // TODO: Add Kernel call for nvptx
+        auto Builder = context->Builder;
+        llvm::Function* parent = Builder->GetInsertBlock()->getParent();
+        std::string opName = AST::UNA_OP_ENUM_STRING[this->op];
+
+        llvm::BasicBlock* addBB = llvm::BasicBlock::Create(Builder->getContext(), opName + ".loop", parent);
+        llvm::BasicBlock* endBB = llvm::BasicBlock::Create(Builder->getContext(), opName + ".done");
+
+        auto indexAlloca = Utils::CreateEntryBlockAlloca(*Builder, "", llvm::Type::getInt64Ty(Builder->getContext()));
+        auto* matSize = Utils::getLength(context, operand, *operandType);
+        auto* nsize = Utils::getLength(context, matAlloc, *operandType);
+        // parent->getBasicBlockList().push_back(addBB);
+        Builder->CreateBr(addBB);
 
-    for (int i = 0; i < dimension.at(0); i++) {
-        // If we have more than one dimension, we need to explore the matrix
-        // more
-        if (dimension.size() > 1) {
-            // Create a new dimension vector with this dimension removed
-            std::vector<int> subDimension(dimension.begin() + 1, dimension.end());
-            recursiveUnaryGeneration(op, module, Builder, ty, matAlloc, opVal, subDimension, (index * prevDim) + i,
-                                     dimension.at(0));
-        } else {
-            // At this point, the element that will be contained is the most raw
-            // llvm value, indexed at position
-            // TODO: Offset needs to work with non-64 bit variables
-            auto zero = llvm::ConstantInt::get(module->getContext(), llvm::APInt(64, 0, true));
-            auto indexVal = llvm::ConstantInt::get(module->getContext(), llvm::APInt(64, index, true));
-            // Pointer to the index within IR
-            auto ptrOld =
-                llvm::GetElementPtrInst::Create(matType, opVal, {zero, indexVal}, "", Builder->GetInsertBlock());
-            auto ptrNew =
-                llvm::GetElementPtrInst::Create(matType, matAlloc, {zero, indexVal}, "", Builder->GetInsertBlock());
-            // Generate the code for each valid operation and type
-            /*TODO: Probably needs syntax checking, leaving this for someone
-             * with a better understanding of programming language theory
-             * */
-            switch (this->op) {
+        Builder->SetInsertPoint(addBB);
+        {
+            auto* index = Builder->CreateLoad(indexAlloca);
+
+            auto* lindex = Builder->CreateURem(index, matSize);
+            auto* v = Utils::getValueFromMatrixPtr(context, operand, lindex, "lhs");
+
+            llvm::Value* opResult;
+            switch (op) {
```

Comment:
This is missing the BNOT case. Is this deliberate?

Reply:
We need to find a way to do binary not, don't think there is an LLVM function to do binary NOT.  This is deliberate until we have a way to do this. Should have mentioned within the meeting.